File: payments/src/customer.rs
use stripe::CustomerId;

use crate::payments_v1::customer_handler_server::*;
use crate::payments_v1::*;

pub struct Customer {
    pub client: stripe::Client,
}

#[async_trait::async_trait]
impl CustomerHandler for Customer {
    async fn create_customer(
        &self,
        request: tonic::Request<CustomerCreateRequest>,
    ) -> Result<tonic::Response<CustomerCreateReply>, tonic::Status> {
        let request = request.into_inner();

        let mut customer = stripe::CreateCustomer::new();
        customer.name = Some(&request.customer_name);
        customer.email = Some(&request.customer_email);

        let result = stripe::Customer::create(&self.client, customer).await;
        if let Ok(customer) = result {
            let reply = CustomerCreateReply {
                customer_id: customer.id.to_string(),
            };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }

    async fn get_customer(
        &self,
        request: tonic::Request<CustomerGetRequest>,
    ) -> Result<tonic::Response<CustomerGetReply>, tonic::Status> {
        let request = request.into_inner();

        let parsed_id = request.customer_id.parse::<CustomerId>();
        if parsed_id.is_err() {
            let e = format!("Invalid customer id: {}", request.customer_id);
            return Err(tonic::Status::new(tonic::Code::InvalidArgument, e));
        }

        let result = stripe::Customer::retrieve(&self.client, &parsed_id.unwrap(), &vec![]).await;
        if let Ok(customer) = result {
            let reply = CustomerGetReply {
                customer_name: customer.name.unwrap_or("".to_string()),
                customer_email: customer.email.unwrap_or("".to_string()),
            };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }

    async fn delete_customer(
        &self,
        request: tonic::Request<CustomerDeleteRequest>,
    ) -> Result<tonic::Response<CustomerDeleteReply>, tonic::Status> {
        let request = request.into_inner();

        let parsed_id = request.customer_id.parse::<CustomerId>();
        if parsed_id.is_err() {
            let e = format!("Invalid customer id: {}", request.customer_id);
            return Err(tonic::Status::new(tonic::Code::InvalidArgument, e));
        }

        let result = stripe::Customer::delete(&self.client, &parsed_id.unwrap()).await;
        if let Ok(_) = result {
            let reply = CustomerDeleteReply { deleted: true };
            Ok(tonic::Response::new(reply))
        } else {
            let e = format!(
                "Internal server error - {}",
                result.unwrap_err().to_string()
            );
            Err(tonic::Status::new(tonic::Code::Internal, e))
        }
    }
}

File: stripe/src/customer.go
package main

import (
	"context"

	pb "github.com/lee-aaron/stripe-go/customer"
	"github.com/stripe/stripe-go/v72"
	"github.com/stripe/stripe-go/v72/customer"
	"google.golang.org/grpc"
)

type CustomerServer struct {
	pb.UnimplementedCustomerHandlerServer
}

func (s *CustomerServer) Register(srv *grpc.Server, cs *CustomerServer) {
	pb.RegisterCustomerHandlerServer(srv, cs)
}

func (s *CustomerServer) CreateCustomer(ctx context.Context, req *pb.CustomerCreateRequest) (*pb.CustomerCreateReply, error) {

	params := &stripe.CustomerParams{
		Name:  &req.CustomerName,
		Email: &req.CustomerEmail,
	}

	for k, v := range req.Metadata {
		params.AddMetadata(k, v)
	}

	c, err := customer.New(params)
	if err != nil {
		return nil, err
	}

	return &pb.CustomerCreateReply{
		CustomerId: c.ID,
	}, nil
}

func (s *CustomerServer) GetCustomer(ctx context.Context, req *pb.CustomerGetRequest) (*pb.CustomerGetReply, error) {

	c, err := customer.Get(req.CustomerId, nil)

	return &pb.CustomerGetReply{
		CustomerName:  c.Name,
		CustomerEmail: c.Email,
		Metadata:      c.Metadata,
	}, err
}

func (s *CustomerServer) DeleteCustomer(ctx context.Context, req *pb.CustomerDeleteRequest) (*pb.CustomerDeleteReply, error) {

	c, err := customer.Del(req.CustomerId, nil)

	return &pb.CustomerDeleteReply{
		Deleted: c.Deleted,
	}, err
}
